packages/components/src/popover/index.js

```diff
@@ -477,32 +478,39 @@ const Popover = (
 		__unstableBoundaryParent,
 	] );
 
+	// Event handlers for closing the popover.
+	const closeEventRef = useRefEffect(
+		( node ) => {
+			function maybeClose( event ) {
+				// Close on escape.
+				if ( event.keyCode === ESCAPE && onClose ) {
+					event.stopPropagation();
+					onClose();
+				}
+			}
+
+			node.addEventListener( 'keydown', maybeClose );
+
+			return () => {
+				node.removeEventListener( 'keydown', maybeClose );
+			};
+		},
+		[ onClose ]
```

Comment:
Nit pick: when we attach event handlers in a ref callback, we usually try to avoid reattaching them when the callback function changes and instead store the callback in a ref and keep it fresh (à la `useLatestRef`) 🙂

Reply:
The main ref callback never changes, but the callback passed to `useRefEffect` will be called every time `onClose` changes.

> Yep, but I don't think `onClose` would change often?

It depends. It's generally a nicer API if you can pass a new function on every render, just as React allows for event handlers. This is also what people are used to. It's easy to forget wrapping a callback in `useCallback`. I believe it's more dev friendly to allow callbacks to change on every render and handle that well.

Reply:
> but the callback passed to useRefEffect will be called every time onClose changes.

Will it? I believe it will just create a new callback function every time `onClose` changes, but the actual ref callback will only be called if the merged ref callback changes. Not tested though, I might have missed something.

Wrapping every callback into `useCallback` is not always a good practice though, memoization also has a cost. Sure, it could be a good default, but I doubt it without further audits. I like to think that React already provides the most general defaults to us, and we shouldn't try to prematurely optimize things. Especially as we already handled it once in `useMergeRefs`.

(I know it's a nit pick, but I also find it interesting 😆)

Reply:
Oh, you're right! I missed that part! Thanks!

I agree. Let's monitor the performance impact (if there's any) before we optimize it. 👍